### PR TITLE
fix: future-proof deserialization of studies with exceptions

### PR DESCRIPTION
**Pull request by hig-dev**
This fixes https://github.com/hpi-studyu/studyu/issues/564

***Notes***

This is an alternative PR to https://github.com/hpi-studyu/studyu/pull/573

It raises exceptions and handles them instead.

### CODE REVIEW

**Comment by johannesvedder** [comment]
Let's merge https://github.com/hpi-studyu/studyu/pull/583 first, then we can merge this one and close https://github.com/hpi-studyu/studyu/pull/573

**Review by johannesvedder — CHANGES_REQUESTED** [review]
I'm encountering the following error with the current version when using the App:

```
1 studies could not be extracted.

======== Exception caught by widgets library =======================================================
The following assertion was thrown building FutureBuilder<ExtractionResult<Study>>(dirty, dependencies: [_LocalizationsScope-[GlobalKey#a7941], _ScaffoldMessengerScope], state: _FutureBuilderState<ExtractionResult<Study>>#f7508):
The showSnackBar() method cannot be called during build.

The showSnackBar() method was called during build, which is prohibited as showing snack bars requires updating state. Updating state is not possible during build.

Instead of calling showSnackBar() during build, call it directly in your on tap (and related) callbacks. If you need to immediately show a snack bar, make the call in initState() or didChangeDependencies() instead. Otherwise, you can also schedule a post-frame callback using SchedulerBinding.addPostFrameCallback to show the snack bar after the current frame.

The ownership chain for the particular ScaffoldMessenger is: "ScaffoldMessenger ← Builder ← DefaultTextStyle ← ValueListenableBuilder<bool> ← CustomPaint ← Banner ← CheckedModeBanner ← Title ← Directionality ← _LocalizationsScope-[GlobalKey#a7941] ← ⋯"
The relevant error-causing widget was: 
  FutureBuilder<ExtractionResult<Study>> FutureBuilder:studyu/flutter_common/lib/src/utils/retry_future_builder.dart:86:12
When the exception was thrown, this was the stack: 
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/errors.dart 297:3       throw_
packages/flutter/src/material/scaffold.dart 347:13                                <fn>
packages/flutter/src/material/scaffold.dart 350:16                                showSnackBar
packages/studyu_app/screens/study/onboarding/study_selection.dart 100:51          <fn>
packages/studyu_flutter_common/src/utils/retry_future_builder.dart 100:27         <fn>
packages/flutter/src/widgets/async.dart 609:48                                    build
packages/flutter/src/widgets/framework.dart 5592:27                               build
packages/flutter/src/widgets/framework.dart 5480:15                               performRebuild
packages/flutter/src/widgets/framework.dart 5643:11                               performRebuild
packages/flutter/src/widgets/framework.dart 5196:7                                rebuild
packages/flutter/src/widgets/framework.dart 2904:18                               buildScope
packages/flutter/src/widgets/binding.dart 989:9                                   drawFrame
packages/flutter/src/rendering/binding.dart 448:5                                 [_handlePersistentFrameCallback]
packages/flutter/src/scheduler/binding.dart 1386:7                                [_invokeFrameCallback]
packages/flutter/src/scheduler/binding.dart 1311:9                                handleDrawFrame
packages/flutter/src/scheduler/binding.dart 1169:5                                [_handleDrawFrame]
lib/_engine/engine/platform_dispatcher.dart 1346:5                                invoke
lib/_engine/engine/platform_dispatcher.dart 260:5                                 invokeOnDrawFrame
lib/_engine/engine/initialization.dart 185:36                                     <fn>
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 550:37  _checkAndCall
dart-sdk/lib/_internal/js_dev_runtime/private/ddc_runtime/operations.dart 555:39  dcall
====================================================================================================

```

File: app/lib/screens/study/onboarding/study_selection.dart [hunk 1]
```diff
@@ -75,18 +91,32 @@ class StudySelectionScreen extends StatelessWidget {
...
+                    final studies = extractionResult!.extracted;
+                    if (extractionResult is ExtractionFailedException<Study>) {
+                      debugPrint('${extractionResult.notExtracted.length} studies could not be extracted.');
+                      WidgetsBinding.instance.addPostFrameCallback((_) {
+                        ScaffoldMessenger.of(context).showSnackBar(
+                          SnackBar(
+                            content: Text(AppLocalizations.of(context)!.study_selection_hidden_studies),
+                          ),
+                        );
+                      });
```
Comment (johannesvedder) [inline app/lib/screens/study/onboarding/study_selection.dart:106]: The snackbar is only shown for a limited amount of time and contains too much information. I would prefer displaying a proper warning to the user instead in form of a `MaterialBanner` between the headline and the `ListView`. The following should work:
```
MaterialBanner(
  leading: Icon(
    MdiIcons.exclamationThick,
    color: Colors.orange,
    size: 32,
  ),
  content: Text(
      AppLocalizations.of(context)!
          .study_selection_hidden_studies,
      style: Theme.of(context).textTheme.titleMedium),
  actions: [Container()],
  backgroundColor: Colors.yellow[100],
);
```

Adapted from [here](https://github.com/hpi-studyu/studyu/blob/d4bc7b5ee7494229bd09c50a6562fd9a53a922b1/app/lib/screens/study/onboarding/eligibility_screen.dart#L111).

**Review by johannesvedder — CHANGES_REQUESTED** [review]
Let's just replace the Snackbar with a Banner. Otherwise the PR looks good!

File: app/lib/screens/study/onboarding/study_selection.dart (hunk 1 above)
Comment (hig-dev) [reply app/lib/screens/study/onboarding/study_selection.dart]: Done

**Review by johannesvedder — CHANGES_REQUESTED** [review]

File: app/lib/screens/study/onboarding/study_selection.dart [hunk 2]
```diff
@@ -74,19 +97,48 @@ class StudySelectionScreen extends StatelessWidget {
...
+                    final studies = extractionResult!.extracted;
+                    if (extractionResult is ExtractionFailedException<Study>) {
+                      debugPrint('${extractionResult.notExtracted.length} studies could not be extracted.');
+                      WidgetsBinding.instance.addPostFrameCallback((_) {
+                        if (_hiddenStudies) return;
+                        setState(() {
+                          _hiddenStudies = true;
+                        });
+                      });
+                    }
```
[the remark this answers is not in the document]
Comment (johannesvedder) [reply app/lib/screens/study/onboarding/study_selection.dart]: I have added a quick fix for this in this PR. Everything else looks fine to me :)

**Review by johannesvedder — APPROVED** [review]

**Pull request merged**